### DISCUSSION TITLE
For some reason, I cannot train Tacotron1 and 2* with A100 ~40GB (google colab) even with batch_size 32/16/8 [tried all 3]

### DISCUSSION THREAD

**Discussion by p0p4k** [coqui-ai/TTS#1460] [General Q&A] [answered]
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/trainer/trainer.py", line 1461, in fit
    self._fit()
  File "/usr/local/lib/python3.7/dist-packages/trainer/trainer.py", line 1445, in _fit
    self.train_epoch()
  File "/usr/local/lib/python3.7/dist-packages/trainer/trainer.py", line 1224, in train_epoch
    _, _ = self.train_step(batch, batch_num_steps, cur_step, loader_start_time)
  File "/usr/local/lib/python3.7/dist-packages/trainer/trainer.py", line 1066, in train_step
    num_optimizers=len(self.optimizer) if isinstance(self.optimizer, list) else 1,
  File "/usr/local/lib/python3.7/dist-packages/trainer/trainer.py", line 946, in _optimize
    outputs, loss_dict = self._model_train_step(batch, model, criterion)
  File "/usr/local/lib/python3.7/dist-packages/trainer/trainer.py", line 902, in _model_train_step
    return model.train_step(*input_args)
  File "/content/drive/MyDrive/coquiTTS/TTS/tts/models/tacotron2.py", line 279, in train_step
    outputs = self.forward(text_input, text_lengths, mel_input, mel_lengths, aux_input)
  File "/content/drive/MyDrive/coquiTTS/TTS/tts/models/tacotron2.py", line 190, in forward
    mel_specs, encoder_outputs, alignments, input_mask
  File "/content/drive/MyDrive/coquiTTS/TTS/tts/models/base_tacotron.py", line 159, in _coarse_decoder_pass
    encoder_outputs.detach(), mel_specs, input_mask
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/content/drive/MyDrive/coquiTTS/TTS/tts/layers/tacotron/tacotron2.py", line 321, in forward
    decoder_output, attention_weights, stop_token = self.decode(memory)
  File "/content/drive/MyDrive/coquiTTS/TTS/tts/layers/tacotron/tacotron2.py", line 273, in decode
    self.context = self.attention(self.query, self.inputs, self.processed_inputs, self.mask)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/content/drive/MyDrive/coquiTTS/TTS/tts/layers/tacotron/attentions.py", line 418, in forward
    torch.tanh(self.static_filter_layer(static_filter) + self.dynamic_filter_layer(dynamic_filter))
RuntimeError: CUDA out of memory. Tried to allocate 20.00 MiB (GPU 0; 39.59 GiB total capacity; 34.26 GiB already allocated; 18.19 MiB free; 37.68 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
An exception has occurred, use %tb to see the full traceback.

**Comment by lexkoro** [comment]
From my experience google colab still keeps the data in the memory even if the run fails. 
So you might wanna check first if it frees up the gpu memory after a failed run.

Also might want to share some more information, like your config and what data you are trying to train on.

**Reply by p0p4k** [reply]
@lexkoro I see. Thanks for your reply.
{
    "output_path": "/content/drive/MyDrive/coquiTTS/notebooks/exp/tacotron",
    "logger_uri": null,
    "run_name": "run",
    "project_name": null,
    "run_description": "\ud83d\udc38Coqui trainer run.",
    "print_step": 25,
    "plot_step": 100,
    "model_param_stats": false,
    "wandb_entity": null,
    "dashboard_logger": "tensorboard",
    "log_model_step": null,
    "save_step": 10000,
    "save_n_checkpoints": 5,
    "save_checkpoints": true,
    "save_all_best": false,
    "save_best_after": 10000,
    "target_loss": null,
    "print_eval": true,
    "test_delay_epochs": -1,
    "run_eval": true,
    "distributed_backend": "nccl",
    "distributed_url": "tcp://localhost:54321",
    "mixed_precision": false,
    "epochs": 1000,
    "batch_size": 4,
    "eval_batch_size": 4,
    "grad_clip": 5.0,
    "scheduler_after_epoch": true,
    "lr": 0.0001,
    "optimizer": "RAdam",
    "optimizer_params": {
        "betas": [
            0.9,
            0.998
        ],
        "weight_decay": 1e-06
    },
    "lr_scheduler": "NoamLR",
    "lr_scheduler_params": {
        "warmup_steps": 4000
    },
    "use_grad_scaler": false,
    "cudnn_enable": true,
    "cudnn_benchmark": true,
    "torch_seed": 54321,
    "model": "tacotron2",
    "num_loader_workers": 4,
    "num_eval_loader_workers": 4,
    "use_noise_augment": false,
    "use_language_weighted_sampler": false,
    "audio": {
        "fft_size": 1024,
        "win_length": 1024,
        "hop_length": 256,
        "frame_shift_ms": null,
        "frame_length_ms": null,
        "stft_pad_mode": "reflect",
        "sample_rate": 22050,
        "resample": true,
        "preemphasis": 0.0,
        "ref_level_db": 20,
        "do_sound_norm": false,
        "log_func": "np.log",
        "do_trim_silence": true,
        "trim_db": 60.0,
        "do_rms_norm": false,
        "db_level": null,
        "power": 1.5,
        "griffin_lim_iters": 60,
        "num_mels": 80,
        "mel_fmin": 0.0,
        "mel_fmax": 8000,
        "spec_gain": 1.0,
        "do_amp_to_db_linear": true,
        "do_amp_to_db_mel": true,
        "pitch_fmax": 640.0,
        "pitch_fmin": 0.0,
        "signal_norm": false,
        "min_level_db": -100,
        "symmetric_norm": true,
        "max_norm": 4.0,
        "clip_norm": true,
        "stats_path": null
    },
    "use_phonemes": false,
    "phonemizer": null,
    "phoneme_language": null,
    "compute_input_seq_cache": false,
    "text_cleaner": "basic_cleaners",
    "enable_eos_bos_chars": false,
    "test_sentences_file": "",
    "phoneme_cache_path": "/content/drive/MyDrive/coquiTTS/notebooks/exp/tacotron/phoneme_cache",
    "characters": {
        "characters_class": "TTS.tts.utils.text.characters.Graphemes",
        "vocab_dict": null,
        "pad": "_",
        "eos": "~",
        "bos": "^",
        "blank": null,
        "characters": "\u1100\u1101\u1102\u1103\u1104\u1105\u1106\u1107\u1108\u1109\u110a\u110b\u110c\u110d\u110e\u110f\u1110\u1111\u1112\u1161\u1162\u1163\u1164\u1165\u1166\u1167\u1168\u1169\u116a\u116b\u116c\u116d\u116e\u116f\u1170\u1171\u1172\u1173\u1174\u1175\u11a8\u11a9\u11aa\u11ab\u11ac\u11ad\u11ae\u11af\u11b0\u11b1\u11b2\u11b3\u11b4\u11b5\u11b6\u11b7\u11b8\u11b9\u11ba\u11bb\u11bc\u11bd\u11be\u11bf\u11c0\u11c1\u11c2",
        "punctuations": "!'(),-.:;? ",
        "phonemes": null,
        "is_unique": true,
        "is_sorted": true
    },
    "add_blank": false,
    "batch_group_size": 0,
    "loss_masking": true,
    "sort_by_audio_len": false,
    "min_audio_len": 1,
    "max_audio_len": Infinity,
    "min_text_len": 1,
    "max_text_len": Infinity,
    "compute_f0": false,
    "compute_linear_spec": false,
    "precompute_num_workers": 0,
    "start_by_longest": false,
    "datasets": [
        {
            "name": "bisang",
            "path": "/content/drive/MyDrive/bisang_audio_data/audio",
            "meta_file_train": "metadata.txt",
            "ignored_speakers": null,
            "language": "",
            "meta_file_val": "",
            "meta_file_attn_mask": ""
        }
    ],
    "test_sentences": [
        "It took me quite a long time to develop a voice, and now that I have it I'm not going to be silent.",
        "Be a voice, not an echo.",
        "I'm sorry Dave. I'm afraid I can't do that.",
        "This cake is great. It's so delicious and moist.",
        "Prior to November 22, 1963."
    ],
    "eval_split_max_size": null,
    "eval_split_size": 0.01,
    "use_gst": false,
    "gst": null,
    "gst_style_input": null,
    "num_speakers": 1,
    "num_chars": 81,
    "r": 2,
    "gradual_training": null,
    "memory_size": -1,
    "prenet_type": "original",
    "prenet_dropout": true,
    "prenet_dropout_at_inference": false,
    "stopnet": true,
    "separate_stopnet": true,
    "stopnet_pos_weight": 10.0,
    "max_decoder_steps": 500,
    "encoder_in_features": 512,
    "decoder_in_features": 512,
    "decoder_output_dim": 80,
    "out_channels": 80,
    "attention_type": "dynamic_convolution",
    "attention_heads": null,
    "attention_norm": "sigmoid",
    "attention_win": false,
    "windowing": false,
    "use_forward_attn": false,
    "forward_attn_mask": false,
    "transition_agent": false,
    "location_attn": true,
    "bidirectional_decoder": false,
    "double_decoder_consistency": true,
    "ddc_r": 6,
    "speakers_file": null,
    "use_speaker_embedding": false,
    "speaker_embedding_dim": 512,
    "use_d_vector_file": false,
    "d_vector_file": false,
    "d_vector_dim": null,
    "seq_len_norm": false,
    "decoder_loss_alpha": 0.25,
    "postnet_loss_alpha": 0.25,
    "postnet_diff_spec_alpha": 0,
    "decoder_diff_spec_alpha": 0,
    "decoder_ssim_alpha": 0,
    "postnet_ssim_alpha": 0,
    "ga_alpha": 0.0
}
Ignore the characters as it is korean characters [i am feeding directly cleaned, phoenemized text instead of plain text]. This is the config that is working for me right now.

**Comment by lexkoro** [comment, marked as answer]
```
"sort_by_audio_len": false,
"min_audio_len": 1,
"max_audio_len": Infinity,
"min_text_len": 1,
"max_text_len": Infinity,
```

So I think you might have some very long samples in your dataset? Which cause the out of memory error.
With max_audio_len and max_text_len you can filter out files which are to long.

When you start training, the max audio and text length of your dataset should be displayed.

Then you could start reducing those values until it fits into the memory ^^.

**Reply by p0p4k** [reply]
Oh, I see!!
Ill sort and try training again. Thanks for the advice!

**Reply by p0p4k** [reply]
I removed all samples more than 20 seconds of audio length. Training tacotron2 at batchsize 32 works perfectly! Thanks again.

**Comment by p0p4k** [comment]
However, my model (training from scratch) did not learn to align even after 10k steps. I will try to train using a pre-trained model and report back.
